### PR TITLE
feat(json): `escape_slash` encoding option

### PR DESCRIPTION
**Pull request by PolyMeilex**
This exposes cjson's `encode_escape_forward_slash ` option to the user, but without global side-effects.
I want to encode a nice human-readable JSON files in my plugin, but escaping every slash (for example in file paths) makes this impossible. 

The only use case I know for forward slash escaping is preventing HTML injections (eg. injecting `</script>` closing tag), so in context of typical nvim plugin it does not matter. (that's why escaping it in JSON is optional)

Probably best reviewed commit by commit:
- `refactor(json): add json encode context type` 
  - Extracts common encoding state to a `json_encode_t` struct (this matches what `json_parse_t` does)
- `feat(json): add call local version of escape_forward_slash`
  - Makes use of the new context struct to implement new `escape_forward_slash` option

### CODE REVIEW

File: test/functional/lua/json_spec.lua [hunk 1]
```diff
@@ -152,6 +152,23 @@ describe('vim.json.encode()', function()
     clear()
   end)
 
+  it('dumps strings with escaped slash', function()
+    eq('"Test\\/"', exec_lua([[return vim.json.encode('Test/')]]))
+    eq('"Test\\/"', exec_lua([[return vim.json.encode('Test/', {})]]))
+    eq('"Test\\/"', exec_lua([[return vim.json.encode('Test/', { __invalid_key = true })]]))
+    eq('"Test\\/"', exec_lua([[return vim.json.encode('Test/', { escape_forward_slash = true })]]))
```
Comment (justinmk) [inline test/functional/lua/json_spec.lua:159]: probably makes sense to also test decode()?

    decode(encode(...))

File: test/functional/lua/json_spec.lua [hunk 2]
```diff
@@ -152,6 +152,23 @@ describe('vim.json.encode()', function()
...
+  end)
+
+  it('dumps strings without escaped slash', function()
+    eq('"Test/"', exec_lua([[return vim.json.encode('Test/', { escape_forward_slash = false })]]))
+    eq(
+      '"Test/"',
+      exec_lua(
+        [[return vim.json.encode('Test/', { __invalid_key = true, escape_forward_slash = false })]]
```
Comment (justinmk) [inline test/functional/lua/json_spec.lua:167]: all these tests should be in 1 test, especially to check that the option does not affect anything globally (and there should be an explicit check for that)

**Comment by bfredl** [comment]
can `escape_forward_slash=false` be the new default? I thought we were encoding JSON, not JSON-embedded-inside-html. Would LSP servers break if we just send them JSON?

File: runtime/doc/lua.txt [hunk 3]
```diff
@@ -812,11 +812,14 @@ vim.json.decode({str}, {opts})                             *vim.json.decode()*
...
 
-vim.json.encode({obj})                                     *vim.json.encode()*
+vim.json.encode({obj}, {opts})                             *vim.json.encode()*
     Encodes (or "packs") Lua object {obj} as JSON in a Lua string.
 
     Parameters: ~
-      • {obj}  (`any`)
+      • {obj}   (`any`)
+      • {opts}  (`table<string,any>?`) Options table with keys:
+                • escape_forward_slash: (boolean) When true, escapes `/`
```
Comment (justinmk) [inline runtime/doc/lua.txt:821]: unnecessarily verbose

```suggestion
                • escape_slash: (boolean) When true, escapes `/`
```

**Comment by justinmk** [comment]
This seems good, and +1 for making it the default (mention this in news.txt).

As a followup, we also need help with merging upstream changes, e.g.: https://github.com/neovim/neovim/pull/24070#issuecomment-1598366646

**Comment by PolyMeilex** [comment]
> can `escape_forward_slash=false` be the new default? I thought we were encoding JSON, not JSON-embedded-inside-html. Would LSP servers break if we just send them JSON?

I would definitely be in favor of that, I have never seen an implementation of JSON that escapes those (except PHP that does it because it was designed to primarily output to HTML). Even JS does not do that.

So, do we keep the option and just set it to `false` by default? Or do we just drop it, and switch it globally?

**Comment by justinmk** [comment]
> So, do we keep the option and just set it to `false` by default?

👍 I didn't look closely, but if you found a good pattern for supporting options in a non-global way, then that seems useful for possibly supporting other options in the future. For reference, we  [removed the cjson global options](https://github.com/neovim/neovim/pull/24070).

**Comment by PolyMeilex** [comment]
> 👍 I didn't look closely, but if you found a good pattern for supporting options in a non-global way, then that seems useful for possibly supporting other options in the future. For reference, we [removed the cjson global options](https://github.com/neovim/neovim/pull/24070).

Yeah, sure, the only reason for this diff being so long is the addition of `json_encode_options_t` that is available in the whole encode call stack for easy addition of new options in the future.

File: runtime/doc/lua.txt (hunk 3 above)
Comment (PolyMeilex) [reply runtime/doc/lua.txt]: Done

File: test/functional/lua/json_spec.lua (hunk 2 above)
Comment (PolyMeilex) [reply test/functional/lua/json_spec.lua]: Merged into 1 test

File: test/functional/lua/json_spec.lua (hunk 1 above)
Comment (PolyMeilex) [reply test/functional/lua/json_spec.lua]: Added two cases for that, decode seems to be up to standard. It decodes `\\/` as `/`.

**Comment by PolyMeilex** [comment]
I believe the comments are addressed, and no escape is now the default.
The failed CI step seems to be a flake?

**Comment by justinmk** [comment]
@PolyMeilex would you mind rebasing this

File: runtime/doc/lua.txt [hunk 4]
```diff
@@ -812,11 +812,14 @@ vim.json.decode({str}, {opts})                             *vim.json.decode()*
...
 
-vim.json.encode({obj})                                     *vim.json.encode()*
+vim.json.encode({obj}, {opts})                             *vim.json.encode()*
     Encodes (or "packs") Lua object {obj} as JSON in a Lua string.
 
     Parameters: ~
-      • {obj}  (`any`)
+      • {obj}   (`any`)
+      • {opts}  (`table<string,any>?`) Options table with keys:
+                • escape_slash: (boolean) When true, escapes `/` character in
```
Comment (justinmk) [inline runtime/doc/lua.txt:821]:
```suggestion
                • escape_slash: (boolean) (default false) When true, escapes `/` character in
```

Comment (PolyMeilex) [reply runtime/doc/lua.txt]: Done

**Pull request merged**